### PR TITLE
New Player property `KeepInvisibilityOnInteracting`.

### PR DESCRIPTION
**Pull request by microsievert**
This property prevents game from removing invisibility effect for player that interacted with something. Added new property and new patch.

### CODE REVIEW

**Review by Ficus-x — APPROVED** [review]

**Comment by NaoUnderscore** [comment]
I don't really find it so useful, as you can use the relative event to deny the original code behavior.
This applies to any properties which replicate the denial behavior of an event.

**Comment by NaoUnderscore** [comment]
As I'm not able to find my previous way to achieve the same result using Exiled API only, I'm gonna consider it as a valid API addition as people may want to prevent players from losing their invisibility status effect regardless their interacting actions.

File: Exiled.API/Features/Player.cs [hunk 1]
```diff
@@ -1083,6 +1083,11 @@ public bool IsInPocketDimension
         /// </summary>
         public bool IsUsingStamina { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets a value indicating whether or not the player invisibility effect should not be removed when interacting.
+        /// </summary>
+        public bool InvisibilityInteractionProtect { get; set; } = false;
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Player.cs:1089]: `KeepInvisibilityOnInteracting` should be better or something.

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Player.cs (hunk 1 above)
Comment (microsievert) [reply Exiled.API/Features/Player.cs]: Solved

File: Exiled.API/Features/Player.cs [hunk 2]
```diff
@@ -1083,6 +1083,11 @@ public bool IsInPocketDimension
         /// </summary>
         public bool IsUsingStamina { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets a value indicating whether or not the player invisibility effect should not be removed when interacting.
+        /// </summary>
+        public bool KeepInvisibilityOnInteracting { get; set; } = false;
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Player.cs:1089]: Just `Gets or sets a value indicating whether the invisibility effect should be removed when interacting.`

Comment (NaoUnderscore) [inline Exiled.API/Features/Player.cs:1089]: `= false` is not needed.

File: Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            for (int z = 0; z < newInstructions.Count; z++)
+                yield return newInstructions[z];
+
+            ListPool<CodeInstruction>.Shared.Return(newInstructions);
+        }
+    }
+
+    /// <summary>
+    /// Patches <see cref="PlayerInteract.OnInteract"/>.
+    /// Implements <see cref="Player.KeepInvisibilityOnInteracting"/> property logic.
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs:71]: Same as above.

File: Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    using Interactables;
+    using Interactables.Verification;
+
+    using NorthwoodLib.Pools;
+
+    using static HarmonyLib.AccessTools;
+
+    /// <summary>
+    /// Patches <see cref="StandardDistanceVerification.ServerCanInteract(ReferenceHub, InteractableCollider)"/>.
+    /// Implements <see cref="Player.KeepInvisibilityOnInteracting"/> property logic.
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs:30]: Remove `property logic.`

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    using NorthwoodLib.Pools;
+
+    using static HarmonyLib.AccessTools;
+
+    /// <summary>
+    /// Patches <see cref="StandardDistanceVerification.ServerCanInteract(ReferenceHub, InteractableCollider)"/>.
+    /// Implements <see cref="Player.KeepInvisibilityOnInteracting"/>.
+    /// </summary>
+    [HarmonyPatch(typeof(StandardDistanceVerification), nameof(StandardDistanceVerification.ServerCanInteract))]
+    internal static class InvisibleInteractionPatch
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs:33]: Rename it as `ServerCanInteractPatch`.

File: Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            ListPool<CodeInstruction>.Shared.Return(newInstructions);
+        }
+    }
+
+    /// <summary>
+    /// Patches <see cref="PlayerInteract.OnInteract"/>.
+    /// Implements <see cref="Player.KeepInvisibilityOnInteracting"/>.
+    /// </summary>
+    [HarmonyPatch(typeof(PlayerInteract), nameof(PlayerInteract.OnInteract))]
+    internal static class InvisibleInteractionPlayerInteractPatch
```
Comment (NaoUnderscore) [inline Exiled.Events/Patches/Generic/InvisibleInteractionPatch.cs:74]: Rename it as `PlayerInteractPatch`.

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Pull request closed without merging**